tests/Functional/ControllerTest.php

```diff
@@ -0,0 +1,211 @@
+    {
+        $this->deleteTempDir();
+    }
+
+    public function testReturnsOne(): void
+    {
+        $request = Request::create('/returnsOne');
+        $request->headers->set('Accept', MimeTypes::APPLICATION_JSON);
+
+        $response = $this->app->handle($request);
+
+        self::assertSame(MimeTypes::APPLICATION_JSON, $response->headers->get('Content-Type'));
+        self::assertJsonStringEqualsJsonString(<<<JSON
+            {"from":"Forest"}
+       JSON, $response->getContent());
+    }
+
+    public function testReturnsMany(): void
+    {
+        $request = Request::create('/returnsMany');
+        $request->headers->set('Accept', MimeTypes::APPLICATION_JSON);
+
+        $response = $this->app->handle($request);
+
+        self::assertSame(MimeTypes::APPLICATION_JSON, $response->headers->get('Content-Type'));
+        self::assertJsonStringEqualsJsonString(<<<JSON
+            [
+                {"from":"Jenny"},
+                {"from":"Forest"}
+            ]
+       JSON, $response->getContent());
+    }
+
+    public function testReconstitutesOneFromRequest(): void
+    {
+        $submitted = <<<JSON
+            {"from":"Jenny"}
+        JSON;
+
+        $request = Request::create('/reconstitutesOne', Request::METHOD_POST, [], [], [], [], $submitted);
+        $request->headers->set('Content-Type', MimeTypes::APPLICATION_JSON);
+        $request->headers->set('Accept', MimeTypes::APPLICATION_JSON);
+
+        $response = $this->app->handle($request);
+
+        self::assertJsonStringEqualsJsonString($submitted, $response->getContent());
+    }
+
+    public function testReconstitutesMultipleFromRequest(): void
+    {
+        $submitted = <<<JSON
+            [
+                {"from":"Jenny"},
+                {"from":"Forest"}
+            ]
+        JSON;
+
+        $request = Request::create('reconstitutesMany', Request::METHOD_POST, [], [], [], [], $submitted);
+        $request->headers->set('Content-Type', MimeTypes::APPLICATION_JSON);
+        $request->headers->set('Accept', MimeTypes::APPLICATION_JSON);
+
+        $response = $this->app->handle($request);
+
+        self::assertJsonStringEqualsJsonString($submitted, $response->getContent());
+    }
+
+    private function deleteTempDir()
+    {
+        if (!file_exists($dir = sys_get_temp_dir() . '/' . Kernel::VERSION . '/CrudControllerTestKernel')) {
+            return;
+        }
+
+        $fs = new Filesystem();
+        $fs->remove($dir);
+    }
+}
+
+/** @internal  */
+final class TestKernel extends Kernel
+{
+    use MicroKernelTrait;
+
+
+    public function __construct()
+    {
+        parent::__construct('test', false);
+    }
+
+    public function registerBundles(): array
+    {
+        return [
+            new FrameworkBundle(),
+            new ViolinesRestBundle(),
+        ];
+    }
+
+    protected function configureRoutes(RouteCollectionBuilder $routes): void
+    {
+        $routes->addRoute(new Route('/returnsOne', [
```

Comment:
looks a bit repetitive. Maybe a first build string HugController::class . '::returnsOne' and then loop over it?

Reply:
Mhh, this is how symfony registers routes. There is some syntactical sugar available when you'd drop support for symfony 4.4.

```php
$routes->add('/returnsOne')->controller([FooController::class, 'foo'])
```

I don't see any benefits in building an abstraction or something here, but instead prefer to do c&p :-)